doc/src/devdocs/locks.md

```diff
@@ -3,158 +3,137 @@
+
+### Level 3
+
+* `jl_method_t.writelock`
+* `typecache_lock`
+* `libmap_lock`
+
+### Level 4
+
+* `jl_methcache_t.writelock`
+
+### Level 5
+
+* `jl_methtable_t.writelock`
+
+### Level 6
+
+* `JuliaOJIT::EmissionMutex` (`std::recursive_mutex`)
+
+* `jl_modules_mutex`
+
+* `jl_uv_mutex` (known as `iolock` from Julia)
+  !!! danger
+      Doing any I/O (for example, printing warning messages or debug information)
+      while holding any other lock listed above may result in pernicious and
+      hard-to-find deadlocks.
+
+* Individual `ThreadSynchronizer` locks
+  !!! danger
+      This may continue to be held after releasing the iolock, or acquired
+      without it, but be very careful to never attempt to acquire the iolock
+      while holding it.
+
+* `Libdl.LazyLibrary.lock` (`ReentrantLock`)
+
+* `orc::ThreadSafeContext`
+  !!! note
+      Acquiring a TSCtx should only be done from the JIT's pool of TSCtx's, and
+      all locks on that TSCtx should be released prior to returning it to the
+      pool. If multiple TSCtx locks must be acquired at the same time (due to
+      recursive compilation), then locks should be acquired in the order that
+      the TSCtxs were borrowed from the pool.
 
-## Locks
-
-Below are all of the locks that exist in the system and the mechanisms for using them that avoid
-the potential for deadlocks (no Ostrich algorithm allowed here):
-
-The following are definitely leaf locks (level 1), and must not try to acquire any other lock:
-
->   * safepoint
->
->     > Note that this lock is acquired implicitly by `JL_LOCK` and `JL_UNLOCK`. use the `_NOGC` variants
->     > to avoid that for level 1 locks.
->     >
->     > While holding this lock, the code must not do any allocation or hit any safepoints. Note that
->     > there are safepoints when doing allocation, enabling / disabling GC, entering / restoring exception
->     > frames, and taking / releasing locks.
->   * shared_map
->   * finalizers
->   * pagealloc
->   * gc_perm_lock
->   * flisp
->   * jl_in_stackwalk (Win32)
->   * ResourcePool<?>::mutex
->   * RLST_mutex
->   * llvm_printing_mutex
->   * jl_locked_stream::mutex
->   * debuginfo_asyncsafe
->   * inference_timing_mutex
->   * ExecutionEngine::SessionLock
->
->     > flisp itself is already threadsafe, this lock only protects the `jl_ast_context_list_t` pool
->     > likewise, the ResourcePool<?>::mutexes just protect the associated resource pool
-
-The following is a leaf lock (level 2), and only acquires level 1 locks (safepoint) internally:
-
->   * global_roots_lock
->   * Module->lock
->   * JLDebuginfoPlugin::PluginMutex
->   * newly_inferred_mutex
-
-The following is a level 3 lock, which can only acquire level 1 or level 2 locks internally:
-
->   * Method->writelock
->   * typecache
-
-The following is a level 4 lock, which can only recurse to acquire level 1, 2, or 3 locks:
-
->   * MethodTable->writelock
-
-No Julia code may be called while holding a lock above this point.
```

Comment:
Preserve this comment?

Reply:
I originally removed it because I interpreted it as "it is safe to call Julia code while holding a lock above level 4", but "it is never safe to call Julia code while holding a level 4 (or lower) lock" makes more sense. Later I'd like to add a little section about what dynamic properties must be true about your Julia code in order to safely hold certain locks, kind of like what we do in `invalidation.jl`.